### ISSUE TITLE
Source Built r1.4 on GPUs with CPU optimized is always slower than 'no cpu optimization'

### ISSUE THREAD

**Issue by inscite** [tensorflow/tensorflow#15725]
### System information
- **Have I written custom code**: Yes, the model named PSIque [arXiv:1711.10644](https://arxiv.org/abs/1711.10644)
- **OS Platform and Distribution**: CentOS 7.1/CentOS 7.3
- **TensorFlow installed from**: source build w/ Bazel
- **TensorFlow version**: 1.4
- **Python version**: Anaconda 3.6.2
- **Bazel version**: 0.8.1
- **GCC/Compiler version (if compiling from source)**: gcc version 4.8.3 20140911 (Red Hat4.8.3-9) (GCC)
- **CUDA/cuDNN version**: CUDA 8.0/r375.26/cuDNN 6.0.0 & CUDA 9.0/r384.81/cuDNN 7.0.5 
- **GPU model and memory**: E5-2660v3*2 Socket, K40m 12GB, P100-PCIE-16GB
- **Exact command to reproduce**: python model.py

### Describe the problem
* I built tensorflow from source for boosting operation performance.

* 6 different distributions were built;
  * CPU only & No CPU optimization (NO EXTRA flags)
  * CPU only & CPU optimization (--config=opt)
  * GPU support & CUDA 8/9 & No CPU optimization (--config=cuda)
  * GPU support & CUDA 8/9 & CPU optimization (--config=opt --config=cuda)

* Experiments were basically done by 5 phases; experiments on CPU only are still going on,
so please focus on GPU version results.

* Results are quite frustrating me, because 'most of CPU optimized versions' gave me slow results.
![results](https://user-images.githubusercontent.com/20734988/34451067-1d8c1cf0-ed5e-11e7-9a4f-3ff21a5c9aea.png)

* Test were made on multiple machine with random order.
  * P100: 2 nodes
  * K40m: 7 nodes
  * CPU only: 8 nodes

* I am curious why CPU optimized version is slow
  * on every experiment combinations
  * even different GPU environments
  * even Dual CPU socket (E5-2660v3)

* (Extra) I believe my current model does not require high throughputs

### tf_env_collect.sh
== cat /etc/issue ===============================================
Linux <hostname> 3.10.0-229.el7.x86_64 #1 SMP Fri Mar 6 11:36:42 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
VERSION="7 (Core)"
VERSION_ID="7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

== are we in docker =============================================
No

== compiler =====================================================
c++ (GCC) 4.8.3 20140911 (Red Hat 4.8.3-9)
Copyright (C) 2013 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

== uname -a =====================================================
Linux vis5 3.10.0-229.el7.x86_64 #1 SMP Fri Mar 6 11:36:42 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

== check pips ===================================================
numpy (1.13.3)
protobuf (3.4.0)
tensorflow (1.4.0)
tensorflow-tensorboard (0.4.0rc3)

== check for virtualenv =========================================
False

== tensorflow import ============================================
tf.VERSION = 1.4.0
tf.GIT_VERSION = b'unknown'
tf.COMPILER_VERSION = b'unknown'
Sanity check: array([1], dtype=int32)

== env ==========================================================
LD_LIBRARY_PATH :/usr/local/cuda/lib64:/usr/local/cuda/extras/CUPTI/lib64
DYLD_LIBRARY_PATH is unset

== nvidia-smi ===================================================
Sat Dec 30 12:39:08 2017
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 384.81                 Driver Version: 384.81                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  On   | 00000000:04:00.0 Off |                    0 |
| N/A   28C    P0    35W / 250W |      0MiB / 16276MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla P100-PCIE...  On   | 00000000:82:00.0 Off |                    0 |
| N/A   35C    P0    38W / 250W |  15661MiB / 16276MiB |     19%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    1     68169      C   python                                     15643MiB |
+-----------------------------------------------------------------------------+

== cuda libs  ===================================================
/usr/local/cuda-8.0/lib64/libcudart_static.a
/usr/local/cuda-8.0/lib64/libcudart.so.8.0.61
/usr/local/cuda-8.0/doc/man/man7/libcudart.so.7
/usr/local/cuda-8.0/doc/man/man7/libcudart.7
/usr/local/cuda-9.0/doc/man/man7/libcudart.7
/usr/local/cuda-9.0/doc/man/man7/libcudart.so.7
/usr/local/cuda-9.0/lib64/libcudart.so.9.0.176
/usr/local/cuda-9.0/lib64/libcudart_static.a

**Comment by bignamehyp** [comment]
Can you please reproduce this result with some of our tutorial/performance code? For example,
https://github.com/tensorflow/benchmarks/tree/master/scripts/tf_cnn_benchmarks

**Comment by inscite** [comment]
Great! I'll have same 5-phase test and share the results. Thx for sharing a benchmark tool.

**Comment by inscite** [comment]
Here's the result!
![tf_cnn_benchmark](https://user-images.githubusercontent.com/20734988/34561299-b0adc1e6-f18c-11e7-956f-dfafc345f287.png)

Since the 'tf-cnn-benchmark' requires nightly version and built version does not compatible with our system, I started from source build (and I believe this version might be dev. version of r1.5).
TensorFlow branch: [6b6d843](https://github.com/tensorflow/tensorflow/tree/6b6d843ccab78f9f91c3b98a43ca09ffecad4747)
Benchmark branch: [ecf4e0b](https://github.com/tensorflow/benchmarks/tree/ecf4e0b474f6f889c6d97b924b0c469decd4f4f9)

- Model: ResNet-50
- NumGpus: 1
- Batch Size: 128
- Batches: 100

As far as I know, it's hard to find difference between cpu w/ opt and w/o optimization.
So CPU optimization is mandatory or not?
I hope this results are helpful to you @bignamehyp !

**Comment by tensorflowbutler** [comment]
The original poster has replied to this issue after the stat:awaiting response label was applied.

**Comment by reedwm** [comment]
So it seems you could not reproduce the problem on `tf_cnn_benchmarks`, but you could PSIque. I'm having trouble reading the table: which columns correspond to Cuda 8 and Cuda 9 and CPU only?

@tfboyd any ideas what the problem could be, or who else could have any ideas?

**Comment by tfboyd** [comment]
My testing with AVX vs SSE3 was on K80 and found the following:

- For synthetic data tests that is almost no difference between SSE3 and AVX2 until 8 GPUs and usually more noticeable with real data. 
- For single GPU, for the models tested.  VGG16, ResNet50, AlexNet, and InceptionV3 the difference was less than 2% until 8 GPUs.
- The biggest difference was AlexNet with SSE3 being 15-18% slower on 8xK80 on a p2.8xlarge AWS instance.  

The results may be different today with improvements to the data pipeline.  When doing GPU training the CPU is mostly only handling the input pipeline.  The input pipeline can benefit from having code optimized for the CPU but it may not be visible.  In the testing above my hypothesis was that because AlexNet pushes the input pipeline the difference becomes visible.  That theory may not be perfect.  TensorFlow will go AVX by default in 1.6 and likely in nightlies in a day or so.  I will try to remember to report back if I see a bump for SSE3 to AVX for the some nightly tests.
